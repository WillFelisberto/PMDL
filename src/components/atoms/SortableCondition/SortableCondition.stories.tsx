import type { Meta, StoryObj } from '@storybook/react';

import { SortableCondition } from './SortableCondition';

const meta: Meta<typeof SortableCondition> = {
  title: 'Atoms/SortableCondition',
  component: SortableCondition,
  tags: ['autodocs'],
  args: {
    children: 'Example',
    variant: 'default'
  }
};

export default meta;

type Story = StoryObj<typeof SortableCondition>;

export const Default: Story = {};
