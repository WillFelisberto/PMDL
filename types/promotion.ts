export type ConditionType =
  | 'comparator'
  | 'iterator'
  | 'group-iterator'
  | 'quantifier'
  | 'operator'
  | 'operation';

export type ConditionBase = {
  id: string;
  type: ConditionType;
  attributes?: Array<{ name: string; value: string }>;
};

export type ComparatorCondition = ConditionBase & {
  type: 'comparator';
  comparatorType:
    | 'equals'
    | 'not-equals'
    | 'less-than'
    | 'less-than-or-equals'
    | 'greater-than'
    | 'greater-than-or-equals'
    | 'contains'
    | 'starts-with'
    | 'ends-with'
    | 'isoneof'
    | 'isnotoneof'
    | 'includes'
    | 'includes-any'
    | 'includes-all';
  leftValue: string;
  rightValue: string;
  dataType: string;
};

export type ValueType = {
  type: 'value' | 'constant' | 'array' | 'null';
  dataType?: string;
  value?: string | number | boolean;
  values?: ValueType[];
};
export type IteratorCondition = ConditionBase & {
  type: 'iterator';
  iteratorType: 'next' | 'every' | 'up-to-and-including' | 'custom';
  collectionName: string;
  elementName: string;
  number?: number;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  condition?: PromotionCondition;
};

export type GroupIteratorCondition = ConditionBase & {
  type: 'group-iterator';
  iteratorType: 'every-group' | 'up-to-and-including-group' | 'custom';
  number?: number;
  conditions?: PromotionCondition[];
};

export type QuantifierCondition = ConditionBase & {
  type: 'quantifier';
  quantifierType: 'exactly' | 'at-least' | 'at-most' | 'all' | 'custom';
  number?: number;
  collectionName: string;
  elementName: string;
  condition?: PromotionCondition;
  customType?: string;
};

export type OperatorCondition = ConditionBase & {
  type: 'operator';
  operatorType: 'and' | 'or' | 'not' | 'custom';
  conditions: PromotionCondition[];
  customType?: string; // Para operadores personalizados
};

export type OperationCondition = ConditionBase & {
  type: 'operation';
  operationType: 'union' | 'anded-union' | 'custom';
  iterators: PromotionCondition[];
  customType?: string;
};
export type PromotionCondition =
  | ComparatorCondition
  | IteratorCondition
  | GroupIteratorCondition
  | QuantifierCondition
  | OperatorCondition
  | OperationCondition;

export type DiscountStructure = {
  calculatorType: string;
  discountType: 'percentOff' | 'amountOff' | 'fixedPrice';
  adjuster: number;
  target?: 'order' | 'shipping' | 'item';
};

export type Offer = {
  discountStructures: DiscountStructure[];
};

export type Promotion = {
  displayName: string;
  description: string;
  enabled: boolean;
  priority: number;
  qualifier: PromotionCondition | undefined;
  offer: Offer;
  conditions: PromotionCondition[];
  sites: Array<{ repositoryId: string }>;
};
