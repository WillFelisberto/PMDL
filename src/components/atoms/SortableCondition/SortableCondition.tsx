import { ComparatorCondition, PromotionCondition } from '@/types/promotion';
import { useSortable } from '@dnd-kit/sortable';
import { CSS } from '@dnd-kit/utilities';
import { GripVertical, X } from 'lucide-react';

export const SortableCondition = ({
  condition,
  onUpdate,
  onDelete
}: {
  condition: PromotionCondition;
  onUpdate: (updated: PromotionCondition) => void;
  onDelete: () => void;
}) => {
  const { attributes, listeners, setNodeRef, transform, transition } = useSortable({
    id: condition.id
  });

  const style = {
    transform: CSS.Transform.toString(transform),
    transition
  };

  if (condition.type !== 'comparator') {
    return (
      <div
        ref={setNodeRef}
        style={style}
        className="group relative bg-gray-100 p-4 rounded-lg border border-gray-300 text-gray-600"
        data-testid="non-editable-condition"
      >
        <p className="text-sm">
          {`Este tipo de condição (${condition.type}) ainda não é editável aqui.`}
        </p>
        <button
          onClick={onDelete}
          data-testid="delete-button"
          className="absolute top-2 right-2 text-red-500 hover:text-red-700"
        >
          <X size={16} />
        </button>
      </div>
    );
  }

  const comparatorCondition = condition as ComparatorCondition;

  return (
    <div
      ref={setNodeRef}
      style={style}
      className="group relative bg-gray-50 p-4 rounded-lg border"
      data-testid="sortable-condition"
    >
      <div className="flex gap-2 items-center">
        <button {...listeners} {...attributes} className="cursor-grab" data-testid="drag-handle">
          <GripVertical className="text-gray-400" />
        </button>

        <select
          className="flex-1 p-2 border rounded-md"
          value={comparatorCondition.comparatorType}
          onChange={(e) =>
            onUpdate({
              ...comparatorCondition,
              comparatorType: e.target.value as ComparatorCondition['comparatorType']
            })
          }
          data-testid="comparator-select"
        >
          <option value="equals">=</option>
          <option value="not-equals">≠</option>
          <option value="greater-than">&gt;</option>
          <option value="greater-than-or-equals">≥</option>
          <option value="less-than">&lt;</option>
          <option value="less-than-or-equals">≤</option>
          <option value="contains">Contém</option>
          <option value="starts-with">Começa com</option>
          <option value="ends-with">Termina com</option>
          <option value="isoneof">É um de</option>
          <option value="isnotoneof">Não é um de</option>
          <option value="includes">Inclui</option>
          <option value="includes-any">Inclui qualquer</option>
          <option value="includes-all">Inclui todos</option>
        </select>

        <input
          className="flex-1 p-2 border rounded-md"
          placeholder="Campo (ex: order.total)"
          value={comparatorCondition.leftValue}
          onChange={(e) => onUpdate({ ...comparatorCondition, leftValue: e.target.value })}
          data-testid="left-value-input"
        />

        <input
          className="flex-1 p-2 border rounded-md"
          placeholder="Valor"
          value={comparatorCondition.rightValue}
          onChange={(e) =>
            onUpdate({
              ...comparatorCondition,
              rightValue: e.target.value
            })
          }
          data-testid="right-value-input"
        />

        <button
          onClick={onDelete}
          className="text-red-500 hover:text-red-700"
          data-testid="delete-button"
        >
          <X size={16} />
        </button>
      </div>
    </div>
  );
};
