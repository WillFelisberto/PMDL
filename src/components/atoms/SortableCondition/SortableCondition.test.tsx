import { SortableCondition } from './SortableCondition';

import { fireEvent, render, screen } from '@/tests/test-utils';
import { ComparatorCondition, PromotionCondition } from '@/types/promotion';

describe('SortableCondition', () => {
  let mockCondition: ComparatorCondition;
  let mockOnUpdate: jest.Mock;
  let mockOnDelete: jest.Mock;

  beforeEach(() => {
    mockCondition = {
      id: 'condition-1',
      type: 'comparator',
      comparatorType: 'equals',
      leftValue: 'order.total',
      rightValue: '100',
      dataType: 'number'
    };

    mockOnUpdate = jest.fn();
    mockOnDelete = jest.fn();
  });

  it('deve renderizar corretamente a condição comparadora', () => {
    const { container } = render(
      <SortableCondition
        condition={mockCondition}
        onUpdate={mockOnUpdate}
        onDelete={mockOnDelete}
      />
    );

    expect(screen.getByTestId('sortable-condition')).toBeInTheDocument();
    expect(screen.getByTestId('drag-handle')).toBeInTheDocument();
    expect(screen.getByTestId('comparator-select')).toHaveValue('equals');
    expect(screen.getByTestId('left-value-input')).toHaveValue('order.total');
    expect(screen.getByTestId('right-value-input')).toHaveValue('100');
    expect(screen.getByTestId('delete-button')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('deve atualizar o tipo de comparador ao alterar o select', () => {
    render(
      <SortableCondition
        condition={mockCondition}
        onUpdate={mockOnUpdate}
        onDelete={mockOnDelete}
      />
    );

    const selectElement = screen.getByTestId('comparator-select');
    fireEvent.change(selectElement, { target: { value: 'greater-than' } });

    expect(mockOnUpdate).toHaveBeenCalledWith({
      ...mockCondition,
      comparatorType: 'greater-than'
    });
  });

  it('deve atualizar o leftValue ao alterar o input correspondente', () => {
    render(
      <SortableCondition
        condition={mockCondition}
        onUpdate={mockOnUpdate}
        onDelete={mockOnDelete}
      />
    );

    const leftValueInput = screen.getByTestId('left-value-input');
    fireEvent.change(leftValueInput, { target: { value: 'order.discount' } });

    expect(mockOnUpdate).toHaveBeenCalledWith({
      ...mockCondition,
      leftValue: 'order.discount'
    });
  });

  it('deve atualizar o rightValue ao alterar o input correspondente', () => {
    render(
      <SortableCondition
        condition={mockCondition}
        onUpdate={mockOnUpdate}
        onDelete={mockOnDelete}
      />
    );

    const rightValueInput = screen.getByTestId('right-value-input');
    fireEvent.change(rightValueInput, { target: { value: '50' } });

    expect(mockOnUpdate).toHaveBeenCalledWith({
      ...mockCondition,
      rightValue: '50'
    });
  });

  it('deve chamar a função onDelete ao clicar no botão de excluir', () => {
    render(
      <SortableCondition
        condition={mockCondition}
        onUpdate={mockOnUpdate}
        onDelete={mockOnDelete}
      />
    );

    const deleteButton = screen.getByTestId('delete-button');
    fireEvent.click(deleteButton);

    expect(mockOnDelete).toHaveBeenCalled();
  });

  it('deve renderizar corretamente uma condição não editável', () => {
    const nonEditableCondition: PromotionCondition = {
      id: 'non-editable',
      type: 'unsupported-type'
    } as unknown as PromotionCondition;

    render(
      <SortableCondition
        condition={nonEditableCondition}
        onUpdate={mockOnUpdate}
        onDelete={mockOnDelete}
      />
    );

    expect(screen.getByTestId('non-editable-condition')).toBeInTheDocument();
    expect(
      screen.getByText(/Este tipo de condição \(unsupported-type\) ainda não é editável aqui./i)
    ).toBeInTheDocument();
  });
});
