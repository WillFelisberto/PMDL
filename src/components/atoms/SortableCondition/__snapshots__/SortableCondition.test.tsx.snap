// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableCondition deve renderizar corretamente a condição comparadora 1`] = `
<div>
  <div
    class="group relative bg-gray-50 p-4 rounded-lg border"
    data-testid="sortable-condition"
  >
    <div
      class="flex gap-2 items-center"
    >
      <button
        aria-describedby=""
        aria-disabled="false"
        aria-roledescription="sortable"
        class="cursor-grab"
        data-testid="drag-handle"
        role="button"
        tabindex="0"
      >
        GripVertical Icon
      </button>
      <select
        class="flex-1 p-2 border rounded-md"
        data-testid="comparator-select"
      >
        <option
          value="equals"
        >
          =
        </option>
        <option
          value="not-equals"
        >
          ≠
        </option>
        <option
          value="greater-than"
        >
          &gt;
        </option>
        <option
          value="greater-than-or-equals"
        >
          ≥
        </option>
        <option
          value="less-than"
        >
          &lt;
        </option>
        <option
          value="less-than-or-equals"
        >
          ≤
        </option>
        <option
          value="contains"
        >
          Contém
        </option>
        <option
          value="starts-with"
        >
          Começa com
        </option>
        <option
          value="ends-with"
        >
          Termina com
        </option>
        <option
          value="isoneof"
        >
          É um de
        </option>
        <option
          value="isnotoneof"
        >
          Não é um de
        </option>
        <option
          value="includes"
        >
          Inclui
        </option>
        <option
          value="includes-any"
        >
          Inclui qualquer
        </option>
        <option
          value="includes-all"
        >
          Inclui todos
        </option>
      </select>
      <input
        class="flex-1 p-2 border rounded-md"
        data-testid="left-value-input"
        placeholder="Campo (ex: order.total)"
        value="order.total"
      />
      <input
        class="flex-1 p-2 border rounded-md"
        data-testid="right-value-input"
        placeholder="Valor"
        value="100"
      />
      <button
        class="text-red-500 hover:text-red-700"
        data-testid="delete-button"
      >
        X Icon
      </button>
    </div>
  </div>
</div>
`;
